server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java

```diff
@@ -119,43 +119,28 @@ Pair<Long,Map<TabletFile,DataFileValue>> reserveFilesForScan() {
 
   void returnFilesForScan(Long reservationId) {
 
-    final Set<StoredTabletFile> filesToDelete = new HashSet<>();
-
-    try {
-      synchronized (tablet) {
-        Set<StoredTabletFile> absFilePaths = scanFileReservations.remove(reservationId);
+    synchronized (tablet) {
+      Set<StoredTabletFile> absFilePaths = scanFileReservations.remove(reservationId);
 
-        if (absFilePaths == null) {
-          throw new IllegalArgumentException("Unknown scan reservation id " + reservationId);
-        }
+      if (absFilePaths == null) {
+        throw new IllegalArgumentException("Unknown scan reservation id " + reservationId);
+      }
 
-        boolean notify = false;
-        try {
-          for (StoredTabletFile path : absFilePaths) {
-            long refCount = fileScanReferenceCounts.decrement(path, 1);
-            if (refCount == 0) {
-              if (filesToDeleteAfterScan.remove(path)) {
-                filesToDelete.add(path);
-              }
-              notify = true;
-            } else if (refCount < 0) {
-              throw new IllegalStateException("Scan ref count for " + path + " is " + refCount);
-            }
-          }
-        } finally {
-          if (notify) {
-            tablet.notifyAll();
+      boolean notify = false;
+      try {
+        for (StoredTabletFile path : absFilePaths) {
+          long refCount = fileScanReferenceCounts.decrement(path, 1);
+          if (refCount == 0) {
+            filesToDeleteAfterScan.add(path);
```

Comment:
Do not need to add to `filesToDeleteAfterScan` here.  This set `filesToDeleteAfterScan` represents files that are in use by a scan and were removed from the metadata table by a compaction.  This method `returnFilesForScan` is called when a scan completes, however the file used may not have been removed from the tablet metadata and may still be active.  Only need to initially add files to `filesToDeleteAfterScan` in `removeFilesAfterScan` which is called by the compaction code when it knows a file its deleting from tablet metadata is also in use by a scan.

```suggestion
```

Reply:
Fixed in 3d00bd9